// @whiskeysockets/baileys/lib/Types/Newsletter.js
export var XWAPaths;
(function (XWAPaths) {
    XWAPaths["xwa2_newsletter_create"] = "xwa2_newsletter_create";
    XWAPaths["xwa2_newsletter_subscribers"] = "xwa2_newsletter_subscribers";
    XWAPaths["xwa2_newsletter_view"] = "xwa2_newsletter_view";
    XWAPaths["xwa2_newsletter_metadata"] = "xwa2_newsletter";
    XWAPaths["xwa2_newsletter_admin_count"] = "xwa2_newsletter_admin";
    XWAPaths["xwa2_newsletter_mute_v2"] = "xwa2_newsletter_mute_v2";
    XWAPaths["xwa2_newsletter_unmute_v2"] = "xwa2_newsletter_unmute_v2";
    XWAPaths["xwa2_newsletter_follow"] = "xwa2_newsletter_follow";
    XWAPaths["xwa2_newsletter_unfollow"] = "xwa2_newsletter_unfollow";
    XWAPaths["xwa2_newsletter_change_owner"] = "xwa2_newsletter_change_owner";
    XWAPaths["xwa2_newsletter_demote"] = "xwa2_newsletter_demote";
    XWAPaths["xwa2_newsletter_delete_v2"] = "xwa2_newsletter_delete_v2";
})(XWAPaths || (XWAPaths = {}));
export var QueryIds;
(function (QueryIds) {
    QueryIds["CREATE"] = "8823471724422422";
    QueryIds["UPDATE_METADATA"] = "24250201037901610";
    QueryIds["METADATA"] = "6563316087068696";
    QueryIds["SUBSCRIBERS"] = "9783111038412085";
    QueryIds["FOLLOW"] = "7871414976211147";
    QueryIds["UNFOLLOW"] = "7238632346214362";
    QueryIds["MUTE"] = "29766401636284406";
    QueryIds["UNMUTE"] = "9864994326891137";
    QueryIds["ADMIN_COUNT"] = "7130823597031706";
    QueryIds["CHANGE_OWNER"] = "7341777602580933";
    QueryIds["DEMOTE"] = "6551828931592903";
    QueryIds["DELETE"] = "30062808666639665";
})(QueryIds || (QueryIds = {}));
